# fix: make mouse playable in XWayland
# Game's original mouse movement behavior was ditched due to complexity to fix it for XWayland.
# Currently the character only moves with mouse, it no longer moves in a constant speed that can be changed by mouse.
# X11 environment is not tested, expect problems!
diff --git a/src/gui.cxx b/src/gui.cxx
index 4156ee5..fbbec88 100644
--- a/src/gui.cxx
+++ b/src/gui.cxx
@@ -19,8 +19,13 @@ static void motionfn ( int x, int y )
 {
   mouse_x = x ;
   mouse_y = y ;
-  mouse_dx += mouse_x - 320 ;
-  mouse_dy += mouse_y - 240 ;
+
+  int viewport[4];
+  glGetIntegerv ( GL_VIEWPORT, viewport ) ;
+
+  mouse_dx = mouse_x - viewport[2] / 2 ;
+  mouse_dy = mouse_y - viewport[3] / 2 ;
+
   puMouse ( x, y ) ;
 }
 
@@ -34,8 +39,11 @@ static void mousefn ( int button, int updown, int x, int y )
   else
     mouse_buttons &= ~(1<<button) ;
 
-  mouse_dx += mouse_x - 320 ;
-  mouse_dy += mouse_y - 240 ;
+  int viewport[4];
+  glGetIntegerv ( GL_VIEWPORT, viewport ) ;
+
+  mouse_dx = mouse_x - viewport[2] / 2 ;
+  mouse_dy = mouse_y - viewport[3] / 2 ;
 
   puMouse ( button, updown, x, y ) ;
 
@@ -43,6 +51,18 @@ static void mousefn ( int button, int updown, int x, int y )
     hide_status () ;
 }
 
+static void set_mousetrap ( int onoff )
+{
+  mousetrap = onoff ;
+
+  if ( mousetrap ) {
+    glutSetCursor ( GLUT_CURSOR_NONE ) ;
+  }
+  else {
+    glutSetCursor ( GLUT_CURSOR_INHERIT ) ;
+  }
+}
+
 static void credits_cb ( puObject * )
 {
   hide_status () ;
@@ -74,28 +94,28 @@ static void help_cb ( puObject * )
 #ifdef XMESA_FX_FULLSCREEN
   static int fullscreen = FALSE ;
   static void setFullscreen () { XMesaSetFXmode ( fullscreen ? XMESA_FX_FULLSCREEN : XMESA_FX_WINDOW ) ; } 
-  static void window_off_cb ( puObject * ) { mousetrap = TRUE  ; fullscreen = TRUE  ; setFullscreen () ; } 
-  static void window_on_cb  ( puObject * ) { mousetrap = FALSE ; fullscreen = FALSE ; setFullscreen () ; } 
+  static void window_off_cb ( puObject * ) { set_mousetrap ( TRUE ) ; fullscreen = TRUE  ; setFullscreen () ; } 
+  static void window_on_cb  ( puObject * ) { set_mousetrap ( FALSE ) ; fullscreen = FALSE ; setFullscreen () ; } 
 #else
   static void setFullscreen () {}
-  static void window_off_cb ( puObject * ) { mousetrap = TRUE  ; setFullscreen () ; } 
-  static void window_on_cb  ( puObject * ) { mousetrap = FALSE ; setFullscreen () ; } 
+  static void window_off_cb ( puObject * ) { set_mousetrap ( TRUE )  ; setFullscreen () ; } 
+  static void window_on_cb  ( puObject * ) { set_mousetrap ( FALSE ) ; setFullscreen () ; } 
 #endif
 #else
   static void setFullscreen () {}
-  static void window_off_cb ( puObject * ) { mousetrap = TRUE  ; setFullscreen () ; } 
-  static void window_on_cb  ( puObject * ) { mousetrap = FALSE ; setFullscreen () ; } 
+  static void window_off_cb ( puObject * ) { set_mousetrap ( TRUE )  ; setFullscreen () ; } 
+  static void window_on_cb  ( puObject * ) { set_mousetrap ( FALSE ) ; setFullscreen () ; } 
 #endif
 
-static void mousetrap_off_cb ( puObject * ) { mousetrap = FALSE ; } 
-static void mousetrap_on_cb  ( puObject * ) { mousetrap = TRUE  ; } 
+static void mousetrap_off_cb ( puObject * ) { set_mousetrap ( FALSE ) ; glutSetCursor ( GLUT_CURSOR_INHERIT ) ; } 
+static void mousetrap_on_cb  ( puObject * ) { set_mousetrap ( TRUE )  ; glutSetCursor ( GLUT_CURSOR_NONE )    ; } 
 
 static void music_off_cb     ( puObject * ) { sound->disable_music () ; } 
 static void music_on_cb      ( puObject * ) { sound->enable_music  () ; } 
 static void sfx_off_cb       ( puObject * ) { sound->disable_sfx   () ; } 
 static void sfx_on_cb        ( puObject * ) { sound->enable_sfx    () ; } 
 
-static void joystick_off_cb  ( puObject * ) { prefer_joystick = FALSE ; mousetrap = TRUE ; } 
+static void joystick_off_cb  ( puObject * ) { prefer_joystick = FALSE ; set_mousetrap ( TRUE ) ; } 
 static void joystick_on_cb   ( puObject * ) { prefer_joystick = TRUE  ; } 
 
 static void set_rcmode       ( puObject * ) { tuxState->setRCMode  ( TRUE  ) ; } 
@@ -138,8 +158,10 @@ GUI::GUI ()
 #endif
 #endif
   hidden = TRUE ;
-  mouse_x = 320 ;
-  mouse_y = 240 ;
+  int viewport[4];
+  glGetIntegerv ( GL_VIEWPORT, viewport ) ;
+  mouse_x = viewport[2] / 2 ;
+  mouse_y = viewport[3] / 2 ;
 
   glutMouseFunc         ( mousefn   ) ;
   glutMotionFunc        ( motionfn  ) ;
@@ -180,8 +202,13 @@ void GUI::show ()
   hide_status () ;
   hidden = FALSE ;
 
+  glutSetCursor ( GLUT_CURSOR_INHERIT ) ;
+
+  int viewport[4];
+  glGetIntegerv ( GL_VIEWPORT, viewport ) ;
+
   if ( mousetrap )
-    glutWarpPointer ( 320, 240 ) ;
+    glutWarpPointer ( viewport[2] / 2, viewport[3] / 2 ) ;
 
 #ifdef SOME_BETTER_WAY_TO_DO_THIS
 #ifdef XMESA_FX_FULLSCREEN
@@ -197,6 +224,14 @@ void GUI::hide ()
   hide_status () ;
   puHideCursor () ;
   main_menu_bar -> hide () ;
+
+  if ( mousetrap )
+  {
+    int viewport[4];
+    glGetIntegerv ( GL_VIEWPORT, viewport ) ;
+    glutWarpPointer ( viewport[2] / 2, viewport[3] / 2 ) ;
+    glutSetCursor ( GLUT_CURSOR_NONE ) ;
+  }
 }
 
 void GUI::update ()
@@ -215,16 +250,19 @@ void GUI::update ()
     Don't let the cursor escape from the screen!
   */
 
+  int viewport[4];
+  glGetIntegerv ( GL_VIEWPORT, viewport ) ;
+
   if ( isHidden () )
-    glutWarpPointer ( 320, 240 ) ;
+    glutWarpPointer ( viewport[2] / 2, viewport[3] / 2 ) ;
   else
   {
     int warp = FALSE ;
 
     if ( mouse_x < 1 ) { mouse_x = 2 ; warp = TRUE ; }
-    else if ( mouse_x > 639 ) { mouse_x = 638 ; warp = TRUE ; }
+    else if ( mouse_x > viewport[2] - 1 ) { mouse_x = viewport[2] - 2 ; warp = TRUE ; }
     if ( mouse_y < 1 ) { mouse_y = 2 ; warp = TRUE ; }
-    else if ( mouse_y > 479 ) { mouse_y = 478 ; warp = TRUE ; }
+    else if ( mouse_y > viewport[3] - 1 ) { mouse_y = viewport[3] - 2 ; warp = TRUE ; }
 
     if ( warp && mousetrap )
       glutWarpPointer ( mouse_x, mouse_y ) ;
@@ -241,7 +279,13 @@ void GUI::joystickInput ()
 
   if ( isHidden() && ( joystick->notWorking() || ! prefer_joystick ) )
   {
-    float mscale = 100.0f ;
+    float mscale = 10.0f ;
+
+    int viewport[4];
+    glGetIntegerv ( GL_VIEWPORT, viewport ) ;
+
+    float mscale_x = mscale * (640.0f / (float) viewport[2]) ;
+    float mscale_y = mscale * (480.0f / (float) viewport[3]) ;
 
     /*
       If he just pressed or just released the right
@@ -257,13 +301,13 @@ void GUI::joystickInput ()
     }
 
     ji.buttons = mouse_buttons ;
-    ji.data[0] = (float)mouse_dx / mscale ;
-    ji.data[1] = (float)mouse_dy / mscale ;
+    ji.data[0] = (float)mouse_dx / mscale_x ;
+    ji.data[1] = (float)mouse_dy / mscale_y ;
 
-    if ( ji.data[0] < -1.0 ) { ji.data[0] = -1.0 ; mouse_dx = (int) -mscale ; }
-    if ( ji.data[0] >  1.0 ) { ji.data[0] =  1.0 ; mouse_dx = (int)  mscale ; }
-    if ( ji.data[1] < -1.0 ) { ji.data[1] = -1.0 ; mouse_dy = (int) -mscale ; }
-    if ( ji.data[1] >  1.0 ) { ji.data[1] =  1.0 ; mouse_dy = (int)  mscale ; }
+    if ( ji.data[0] < -1.0 ) { ji.data[0] = -1.0 ; mouse_dx = (int) -mscale_x ; }
+    if ( ji.data[0] >  1.0 ) { ji.data[0] =  1.0 ; mouse_dx = (int)  mscale_x ; }
+    if ( ji.data[1] < -1.0 ) { ji.data[1] = -1.0 ; mouse_dy = (int) -mscale_y ; }
+    if ( ji.data[1] >  1.0 ) { ji.data[1] =  1.0 ; mouse_dy = (int)  mscale_y ; }
 
     if ( ji.data[0] < 0.1 && ji.data[0] > -0.1 ) ji.data[0] =  0.0 ;
     if ( ji.data[1] < 0.1 && ji.data[1] > -0.1 ) ji.data[1] =  0.0 ;
